### PR TITLE
fix build on macos

### PR DESCRIPTION
**Pull request by yjh0502**
fixes https://github.com/kahing/catfs/issues/10

 - most missing syscalls (like `fdopendir`) are fixed by updating `libc` crate, that's why `Cargo.lock` is updated to latest
 - some other syscalls are emulated (like `splice`)
 - type errors (`u16` vs `u32`) are fixed by adding castings.

### CODE REVIEW

**Comment by gaul** [comment]
@yjh0502 Parts of this look good but I made a #32 with a different approach with `mode_t`.  When that merges, breaking this PR into two more, one updating the libc crate and the other emulating syscalls, will make this easier to merge.

**Comment by gaul** [comment]
@yjh0502 can you rebase this?  It will require some work to use the correct libc calls, e.g., the non-64-bit variants.

**Comment by gaul** [comment]
Superseded by #45.

File: src/catfs/rlibc.rs [hunk 1]
```diff
@@ -460,6 +493,11 @@ impl File {
         }
     }
 
+    #[cfg(target_os = "macos")]
+    pub fn allocate(&self, offset: u64, len: usize) -> io::Result<()> {
+        self.truncate((offset as usize)+ len)
```
Comment (Artoria2e5) [inline src/catfs/rlibc.rs:498]: This should ideally be replaced with some `fcntl` stuff. See https://stackoverflow.com/a/11497568/3770260.

But sure, we are all using SSD these days so speed isn't a concern. And Apple products don't really have longivity concerns.

File: src/catfs/rlibc.rs [hunk 2]
```diff
@@ -296,7 +322,10 @@ pub fn fstatat(dir: RawFd, path: &AsRef<Path>) -> io::Result<libc::stat> {
     let stp = (&mut st) as *mut libc::stat;
     let s = to_cstring(path);
 
+    #[cfg(not(target_os = "macos"))]
     let res = unsafe { libc::fstatat(dir, s.as_ptr(), stp, libc::AT_EMPTY_PATH) };
+    #[cfg(target_os = "macos")]
+    let res = unsafe { libc::fstatat(dir, s.as_ptr(), stp, 0) };
```
Comment (Artoria2e5) [inline src/catfs/rlibc.rs:328]: Nope, this won't work for an empty path, which is exactly what `AT_EMPTY_PATH` is meant to do. Just run this C code on a Mac:

```C
#include <stdio.h>
#include <sys/stat.h>
#include <fcntl.h>
int main(void) {
    struct stat buf;
    printf("%d\n", fstatat(AT_FDCWD, "", &buf, 0));
    perror("fstatat");
    return 0;
}
```

A racy substitute might be needed, using a path from [`F_GETPATH` fcntl](https://stackoverflow.com/a/13544447/3770260).

File: src/catfs/rlibc.rs (hunk 1 above)
Comment (gaul) [reply src/catfs/rlibc.rs:498]: This PR is superseded by #45.  Could you review that instead?

Comment (Artoria2e5) [reply src/catfs/rlibc.rs:498]: Sure.

**Comment by gaul** [comment]
Fix via #45.  Please test with the latest master.

**Pull request closed without merging**